acvm-repo/acvm/src/compiler/optimizers/merge_expressions.rs

```diff
@@ -109,17 +127,22 @@ impl MergeExpressionsOptimizer {
                     }
                 }
             }
+        }
+
+        // Construct the new circuit from modified/deleted gates
+        let mut new_circuit = Vec::new();
+        let mut new_acir_opcode_positions = Vec::new();
 
-            if to_keep {
-                let opcode = modified_gates.get(&i).cloned().unwrap_or(opcode);
-                new_circuit.push(opcode);
-                new_acir_opcode_positions.push(opcode_position);
+        for (i, opcode_position) in acir_opcode_positions.iter().enumerate() {
+            if let Some(op) = self.get_opcode(i, circuit) {
+                new_circuit.push(op);
+                new_acir_opcode_positions.push(*opcode_position);
             }
         }
         (new_circuit, new_acir_opcode_positions)
     }
 
-    fn brillig_input_wit<F>(&self, input: &BrilligInputs<F>) -> BTreeSet<Witness> {
+    fn brillig_input_wit(&self, input: &BrilligInputs<F>) -> BTreeSet<Witness> {
```

Comment:
This could also be rewritten in terms of [WitnessFolder::fold_brillig_inputs](https://github.com/noir-lang/noir/blob/70cb16a2d38a1d6dc893371ade88e453ccc99070/acvm-repo/acvm/src/compiler/transformers/mod.rs#L324)